qualtran/simulation/tensor/_tensor_data_manipulation.py

```diff
@@ -0,0 +1,134 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+"""Utility methods to generate and manipulate tensor data for Bloqs."""
+import itertools
+from typing import List, Tuple, Union
+
+import attrs
+import numpy as np
+
+from qualtran import CtrlSpec, Side, Signature
+
+
+def tensor_out_inp_shape_from_signature(
+    signature: Signature,
+) -> Tuple[Tuple[int, ...], Tuple[int, ...]]:
+    """Returns a tuple for tensor data corresponding to signature.
+
+    Tensor data for a bloq with a given `signature` can be expressed as a ndarray of
+    shape `out_indices_shape + inp_indices_shape` where
+
+     1. out_indices_shape - A tuple of values `2 ** soq.reg.bitsize` for every soquet `soq`
+         corresponding to the RIGHT registers in signature.
+     2. inp_indices_shape - A tuple of values `2 ** soq.reg.bitsize` for every soquet `soq`
+         corresponding to the LEFT registers in signature.
+
+    This method returns a tuple of (out_indices_shape, inp_indices_shape).
+    """
+    inp_indices_shape = [2**reg.bitsize for reg in signature.lefts() for _ in reg.all_idxs()]
+    out_indices_shape = [2**reg.bitsize for reg in signature.rights() for _ in reg.all_idxs()]
+    return tuple(out_indices_shape), tuple(inp_indices_shape)
+
+
+def tensor_shape_from_signature(signature: Signature) -> Tuple[int, ...]:
+    """Returns a tuple for tensor data corresponding to signature.
+
+    Tensor data for a bloq with a given `signature` can be expressed as a ndarray of
+    shape `out_indices_shape + inp_indices_shape` where
+
+     1. out_indices_shape - A tuple of values `2 ** soq.reg.bitsize` for every soquet `soq`
+         corresponding to the RIGHT registers in signature.
+     2. inp_indices_shape - A tuple of values `2 ** soq.reg.bitsize` for every soquet `soq`
+         corresponding to the LEFT registers in signature.
+
+    This method returns a tuple of (*out_indices_shape, *inp_indices_shape).
+    """
+    out_shape, inp_shape = tensor_out_inp_shape_from_signature(signature)
+    return out_shape + inp_shape
+
+
+def active_space_for_ctrl_spec(
+    signature: Signature, ctrl_spec: CtrlSpec
+) -> Tuple[Union[int, slice], ...]:
+    """Returns the "active" subspace corresponding to `signature` and `ctrl_spec`.
+
+    Assumes first n-registers for `signature` are control registers corresponding to `ctrl_spec`.
```

Comment:
say somewhere that this is a sequence of indices/slices that can address into a ndarray

Reply:
done